### PR TITLE
🐛   Merge load balancer properties with existing properties to avoid overwriting cloud provider rules

### LINKED ISSUE #1139: azure nodeOutbound loadbalancer reconcile fails after creating serviceType=LoadBalancer
/kind bug

**What steps did you take and what happened:**
Create a cluster using clusterctl 

```
clusterctl version
clusterctl version: &version.Info{Major:"0", Minor:"3", GitVersion:"v0.3.12", GitCommit:"9e1dd7e8e428e05bee406602952ae269d55bdbba", GitTreeState:"clean", BuildDate:"2020-12-15T16:42:40Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"darwin/amd64"}
```

```
clusterctl init --infrastructure=azure
Fetching providers
Installing cert-manager Version="v0.16.1"
Waiting for cert-manager to be available...
Installing Provider="cluster-api" Version="v0.3.13" TargetNamespace="capi-system"
Installing Provider="bootstrap-kubeadm" Version="v0.3.13" TargetNamespace="capi-kubeadm-bootstrap-system"
Installing Provider="control-plane-kubeadm" Version="v0.3.13" TargetNamespace="capi-kubeadm-control-plane-system"
Installing Provider="infrastructure-azure" Version="v0.4.10" TargetNamespace="capz-system"
```

Install cni
[cni.yaml.txt](https://github.com/kubernetes-sigs/cluster-api-provider-azure/files/5868627/cni.yaml.txt)

Install nginx controller 
`kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/controller-v0.43.0/deploy/static/provider/cloud/deploy.yaml`

and capz starts throwing error

**What did you expect to happen:**
Expect no reconcile error in capz
```
I0125 17:50:26.321627       1 azurecluster_controller.go:168] controllers/AzureCluster "msg"="Reconciling AzureCluster" "azureCluster"="<cluster-name>" "cluster"="<cluster-name>" "namespace"="default" 
E0125 17:50:30.964200       1 controller.go:257] controller-runtime/controller "msg"="Reconciler error" "error"="failed to reconcile cluster services: failed to reconcile load balancer: failed to create load balancer \"cluster-name\": network.LoadBalancersClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code=\"InvalidResourceReference\" Message=\"Resource /subscriptions/<subscriptionId>/resourceGroups/<cluster-name>/providers/Microsoft.Network/loadBalancers/ds-bet-1780/backendAddressPools/<cluster-name> referenced by resource /subscriptions/<subscriptionId>/resourceGroups/<cluster-name>/providers/Microsoft.Network/loadBalancers/<subscriptionId>/loadBalancingRules/a285df35053e544419fd8ecb70ac7218-TCP-80 was not found. Please make sure that the referenced resource exists, and that both resources are in the same region.\" Details=[]" "controller"="azurecluster" "name"="<cluster-name>" "namespace"="default"
```

**Anything else you would like to add:**

**Environment:**

- cluster-api-provider-azure version: v0.4.10
- Kubernetes version: (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-14T05:15:04Z", GoVersion:"go1.15.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.8", GitCommit:"9f2892aab98fe339f3bd70e3c470144299398ace", GitTreeState:"clean", BuildDate:"2020-08-13T16:04:18Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```
- OS (e.g. from `/etc/os-release`): 
`Darwin 20.2.0 Darwin Kernel Version 20.2.0: Wed Dec  2 20:39:59 PST 2020; root:xnu-7195.60.75~1/RELEASE_X86_64 x86_64`

### PR DESCRIPTION
**Pull request by CecileRobertMichon**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:  Fixes a bug where cloud provider creates Load Balancing rules and backend pools for services on the "node outbound" LB and CAPZ subsequently tries to reconcile the LB and overwrites the backendPools, but not the LB rules, resulting in an error `InvalidResourceReference` because the LB rule refers to a nonexistent backend pool.

New behavior is:
1) GET the LB
2) merge existing LB properties with desired CAPZ defaults
3) if no update is needed, DONE
4) if an update is needed PUT the LB with an etag (only update if no changes have been made since)

Did some (slight) refactoring of the loadbalancer service to help reuse of code which IMO made the code easier to read (the previous Reconcile func was a litttle hard to follow)...

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1139 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Merge load balancer properties with existing properties to avoid overwriting cloud provider rules
```

### CODE REVIEW

File: cloud/services/loadbalancers/loadbalancers.go [hunk 1]
```diff
@@ -206,5 +215,128 @@ func (s *Service) getFrontendIPConfigs(lbSpec azure.LBSpec) ([]network.FrontendI
...
+					IntervalInSeconds: to.Int32Ptr(15),
+					NumberOfProbes:    to.Int32Ptr(4),
+				},
+			},
+		}
+	}
+	return []network.Probe{}
+}
+
+func probeExists(probes []network.Probe, probe network.Probe) bool {
```
Comment (CecileRobertMichon) [inline cloud/services/loadbalancers/loadbalancers.go:299]: note: these exists helpers look _very_ similar... I tried to combine them into one and use reflect to get the Name field, like this:

```
func exists(existing []interface{}, item interface{}) bool { 
wanted := reflect.ValueOf(&item).Elem().FieldByName("Name").Elem().String() 
for _, v:=range existing {
     name:=reflect.ValueOf(&v).Elem().FieldByName("Name").Elem().String()
     if name == wanted { 
        return true 
    }
} 
return false 
}
```
But then I realized because I have to deal with slices of structs, converting them to `[]interface{}` is costly since it's O(n) so in the end I decided the duplication of code is not worse.

File: cloud/services/loadbalancers/client.go [hunk 2]
```diff
@@ -66,8 +66,22 @@ func (ac *AzureClient) CreateOrUpdate(ctx context.Context, resourceGroupName str
...
+	var etag string
+	if lb.Etag != nil {
+		etag = *lb.Etag
+	}
+	req, err := ac.loadbalancers.CreateOrUpdatePreparer(ctx, resourceGroupName, lbName, lb)
+	if err != nil {
+		err = autorest.NewErrorWithError(err, "network.LoadBalancersClient", "CreateOrUpdate", nil, "Failure preparing request")
+		return err
+	}
+	if etag != "" {
```
Comment (devigned) [inline cloud/services/loadbalancers/client.go:78]: Please consider adding vertical spacing after each of the if blocks.

File: cloud/services/loadbalancers/loadbalancers.go [hunk 3]
```diff
@@ -61,14 +67,79 @@ func (s *Service) Reconcile(ctx context.Context) error {
...
+			// merge existing LB properties with desired properties
+			frontendIPConfigs = *existingLB.FrontendIPConfigurations
+			wantedIPs, wantedFrontendIDs := s.getFrontendIPConfigs(lbSpec)
+			for _, ip := range wantedIPs {
+				if !ipExists(frontendIPConfigs, ip) {
+					update = true
+					frontendIPConfigs = append(frontendIPConfigs, ip)
+				}
+			}
+			loadBalancingRules = *existingLB.LoadBalancingRules
```
Comment (devigned) [inline cloud/services/loadbalancers/loadbalancers.go:98]: Please consider adding vertical spacing after each of the for blocks.

File: cloud/services/loadbalancers/loadbalancers.go [hunk 4]
```diff
@@ -61,14 +67,79 @@ func (s *Service) Reconcile(ctx context.Context) error {
 	defer span.End()
 
 	for _, lbSpec := range s.Scope.LBSpecs() {
-		s.Scope.V(2).Info("creating load balancer", "load balancer", lbSpec.Name)
+		var etag *string
```
Comment (devigned) [inline cloud/services/loadbalancers/loadbalancers.go:70]: wdyt of:
```go
	var (
		etag                *string
		frontendIDs         []network.SubResource
		frontendIPConfigs   = make([]network.FrontendIPConfiguration, 0)
		loadBalancingRules  = make([]network.LoadBalancingRule, 0)
		backendAddressPools = make([]network.BackendAddressPool, 0)
		outboundRules       = make([]network.OutboundRule, 0)
		probes              = make([]network.Probe, 0)
	)
```

**Review by devigned — COMMENTED** [review]
Really like the refactoring work to make the code more readable. Just left a few nits.

**Review by devigned — APPROVED** [review]
/lgtm

/assign @nader-ziada
